scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5082 @@
+            infrastructure chooses between the following options.
+
+            - ``'formula'``: use a formula for the log-PDF itself
+            - ``'logexp'``: evaluate the PDF and takes its logarithm
+
+            Not all `method` options are available for all distributions.
+            If the selected `method` is not available, a ``NotImplementedError``
+            will be raised.
+
+        Returns
+        -------
+        out : Array
+            The log-PDF evaluated at the argument `x`.
+
+        See Also
+        --------
+        ContinuousDistribution.pdf
+        ContinuousDistribution.logcdf
+
+        Notes
+        -----
+        Suppose a probability distribution has support :math:`[l, r]`.
+        By definition of the support, the log-PDF evaluates to its minimum value
+        of :math:`-\infty` (i.e. :math:`\log(0)`) outside the support; i.e. for
+        :math:`x < l` or :math:`x > r`. The maximum of the log-PDF may be less
+        than or greater than :math:`\log(1) = 0` because the maximum of the PDF
+        can be any positive real.
+
+        For distributions with infinite support, it is common for
+        `ContinuousDistribution.pdf` to return a value of `0` when the argument
+        is theoretically within the support; this can occur because the true value
+        of the PDF is too small to be represented by the chosen dtype. The log
+        of the PDF, however, will often be numerically nonzero over a much larger
+        domain. Consequently, it may be preferred to work with the logarithms of
+        probabilities and probability densities to avoid underflow.
+
+        Examples
+        --------
+        Instantiate a distribution with the desired parameters:
+
+        >>> import numpy as np
+        >>> from scipy import stats
+        >>> X = stats.Uniform(a=-1.0, b=1.0)
+
+        Evaluate the log-PDF at the desired argument:
+
+        >>> X.logpdf(0.5)
+        -0.6931471805599453
+        >>> np.allclose(X.logpdf(0.5), np.log(X.pdf(0.5)))
+        True
+
+        """
+        return self._logpdf_dispatch(x, method=method, **self._parameters)
+
+    @_dispatch
+    def _logpdf_dispatch(self, x, *, method=None, **kwargs):
+        if self._overrides('_logpdf_formula'):
+            method = self._logpdf_formula
+        elif self.tol is _null:  # ensure that developers override _logpdf
+            method = self._logpdf_logexp
+        return method
+
+    def _logpdf_formula(self, x, **kwargs):
+        raise NotImplementedError(self._not_implemented)
+
+    def _logpdf_logexp(self, x, **kwargs):
+        return np.log(self._pdf_dispatch(x, **kwargs))
+
+    @_set_invalid_nan
+    def pdf(self, x, *, method=None):
+        r"""Probability density function
+
+        The probability density function :math:`f(x)` is the probability *per
+        unit interval* of :math:`x` that the random variable takes on the value
+        :math:`x`. Mathematically, it is defined as the derivative of the
```

Comment:
note sure how precise we want to be. F might not be differentiable at every x (only at almost every x :). you could add: if F is differentiable, ...

Reply:
When we're done, I'm sure there will be dozens of other occurences where mathematically subtleties have been glossed over. I don't think we need to address the ones that are not relevant to the reader of the class documentation. Are there any SciPy distributions now in which the CDF is not differentiable at points within the support and we return NaN for the PDF? Even if there are, this is not the case in the distributions currently included or that we've discussed adding.

I've changed this from "is defined as" to "can be defined as". The latter is true, because it doesn't suggest that it is *always* defined that way.